src/simulators/statevector/qubitvector_thrust.hpp

```diff
@@ -0,0 +1,3146 @@
+
+#ifdef QASM_HAS_ATS
+	pMat = (thrust::complex<double>*)&state0;
+#else
+	uint_t i,matSize;
+	matSize = 1ull << N;
+
+	allocate_buffers(N);
+
+	pMat = m_pMatDev;
+#pragma omp parallel for 
+	for(i=0;i<matSize*matSize;i++){
+		m_pMatDev[i] = state0[i];
+	}
+#endif
+
+	size = data_size_ >> N;
+	apply_lambda(initialize_component_lambda<data_t>((thrust::complex<data_t>*)&data_[0],pMat,m_pUintBuf,(uint_t*)&qubits[0],N), size);
+}
+
+//------------------------------------------------------------------------------
+// Utility
+//------------------------------------------------------------------------------
+
+template <typename data_t>
+struct fill_lambda : public unary_function<uint_t,void>
+{
+	thrust::complex<data_t>* pVec;
+	thrust::complex<data_t> val;
+
+	fill_lambda(thrust::complex<data_t>* pV,thrust::complex<data_t>& v)
+	{
+		pVec = pV;
+		val = v;
+	}
+
+	__host__ __device__ void operator()(const uint_t &i) const
+	{
+		pVec[i] = val;
+	}
+};
+
+template <typename data_t>
+void QubitVectorThrust<data_t>::zero()
+{
+	uint_t size;
+	thrust::complex<data_t> z = 0.0;
+
+	size = data_size_;
+	apply_lambda(fill_lambda<data_t>((thrust::complex<data_t>*)&data_[0],z), size);
+}
+
+template <typename data_t>
+cvector_t<data_t> QubitVectorThrust<data_t>::convert(const cvector_t<double>& v) const {
+  cvector_t<data_t> ret(v.size());
+  for (size_t i = 0; i < v.size(); ++i)
+    ret[i] = v[i];
+  return ret;
+}
+
+
+template <typename data_t>
+void QubitVectorThrust<data_t>::set_num_qubits(size_t num_qubits) {
+
+  size_t prev_num_qubits = num_qubits_;
+  num_qubits_ = num_qubits;
+  data_size_ = BITS[num_qubits];
+
+  if (checkpoint_) {
+    free(checkpoint_);
+    checkpoint_ = nullptr;
+  }
+
+  // Free any currently assigned memory
+  if (data_) {
+    if (prev_num_qubits != num_qubits_) {
+    	if(m_useATS){
+    		free(data_);
+    	}
+    	else{
+	    	cudaFree(data_);
+    	}
+    	data_ = nullptr;
+    }
+  }
+
+	int tid,nid;
+	char* str;
+
+	nid = omp_get_num_threads();
+	tid = omp_get_thread_num();
+	cudaGetDeviceCount(&m_nDev);
+
+	m_iDev = 0;
+	if(nid > 1){
+		m_iDev = tid % m_nDev;
+		cudaSetDevice(m_iDev);
+		m_nDevParallel = 1;
+	}
+	else{
+		m_nDevParallel = 1;
+		str = getenv("QASM_MULTI_GPU");
+		if(str != NULL){
+			m_nDevParallel = m_nDev;
+		}
+	}
+
+	// Allocate memory for new vector
+	if (data_ == nullptr){
+		void* pData;
+
+#ifdef QASM_TIMING
+		TimeReset();
+		TimeStart(QS_GATE_INIT);
+#endif
+		str = getenv("QASM_USE_ATS");
+		if(str != NULL){
+			posix_memalign(&pData,128,sizeof(thrust::complex<data_t>) * data_size_);
+			m_useATS = 1;
+		}
+		else{
+			str = getenv("QASM_USE_DEVMEM");
+			if(str != NULL){
+				cudaMalloc(&pData,sizeof(thrust::complex<data_t>) * data_size_);
+				m_useDevMem = 1;
+			}
+			else{
+				cudaMallocManaged(&pData,sizeof(thrust::complex<data_t>) * data_size_);
+			}
+			m_useATS = 0;
+		}
+		data_ = reinterpret_cast<std::complex<data_t>*>(pData);
+
+		if(m_nDevParallel > 1){
+			/*
```

Comment:
Is this commented out code block supposed to be here?

Reply:
deleted